### PR TITLE
Implement multiple chain state tracking

### PR DESCRIPTION
**Pull request by yaahc**
This can be collapsed with the previous `chain-reorg` PR if that is preferred.

### CODE REVIEW

**Comment by teor2345** [comment]
> This can be collapsed with the previous `chain-reorg` PR if that is preferred.

I feel like we might need

**Comment by teor2345** [comment]
Oops I hit the wrong button.

File: zebra-state/src/in_memory/chains.rs [hunk 1]
```diff
@@ -3,18 +3,97 @@ use crate::{Request, Response};
...
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+}
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: For the sake of other reviewers:
```suggestion
    block: Arc<Block>,
    // TODO: add more state here
}
```

File: zebra-state/src/in_memory/chains.rs [hunk 2]
```diff
@@ -3,18 +3,97 @@ use crate::{Request, Response};
...
 
 /// A service wrapper that tracks multiple chains, handles reorgs, and persists
 /// blocks to disk once they're past the reorg limit
 pub(crate) struct ChainsState<S> {
     /// The inner state service that only tracks a single chain
     inner: S,
+    /// The set of chains
+    //
+    // might need to use a map type and pop / reinsert with cummulative work as the index
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: The consensus rule is: "the main tip has the highest cumulative work, breaking ties with the first block received by the node".

So I think zebra-consensus should manage the tip choice, but I'd like to get others' input.
```suggestion
    // might need to use a map type and pop / reinsert with (TODO based on the final design)
```

Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs]: Did you intend for the suggestion to make no changes?

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: I fixed it. I think the alternatives are:
* (cumulative work: u64, timestamp: i64)
* is_main_tip: bool
* store the tips as a separate array, and the main tip as a reference, and make the cumulative work and timestamp part of the chain state struct

It's impossible for the ancestor of a tip to have a higher cumulative work than the tip. So I don't think we need to track all the recent blocks as if they are tips.

There's an edge case where:
* there are two equal work tips
* we restart and discard their timestamps
* we reload and have to break the tie

Since the tie-breaker is local to the node, we can do whatever we want here. We could choose the numerically lower hash. Or the one that comes off the disk first.

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: (I added some more design info.)

File: zebra-state/src/in_memory/chains.rs [hunk 3]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+    // TODO: add more state here
+}
+
+/// A persistent data structure representing the end of a chain
+struct VolatileChain(im::OrdMap<BlockHeight, ChainState>);
```
Comment (dconnolly) [inline zebra-state/src/in_memory/chains.rs]: `Chain`, or `ChainSlice` perhaps?

File: zebra-state/src/in_memory/chains.rs (hunk 2 above)
Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs]: @teor2345 I don't know what you mean by "So I don't think we need to track all the recent blocks as if they are tips."

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: I'm not sure if we're thinking about the same design.

How many items are in this set?
Just a few, or about 100?

Is it tracking just the tips, or all the recent blocks?

Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs]: its tracking all recent blocks, and it's approximately 100 but in practice probably larger because I plan on only removing blocks from the chains when they're persisted to disk, so all the blocks from chains that never get finalized will live on until the next restart.

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: How is the inner state service used?
Which blocks does it have in it?
Why does there need to be a reference here?

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: Oh, right, I'm really confused, because there is a ChainState and Chain**s**State. (Notice the extra "s".)

Let's rename them so it's clear what they do, and why they both exist.

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: Oh, and there's a VolatileChain as well, which is "persistent".

Is there a comment somewhere that explains how these 3 data structures are used, and how they relate to each other?

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think this could work, and there are some great ideas here.

But I'm finding it hard to work out the overall structure, and how each struct is used.

I'd like to do another review when we have a design doc, or a detailed module comment. It should explain each struct, its purpose, how it is used, and how they relate to one another.

Once we have that design, it should be easier to choose distinct names for each struct.

File: zebra-state/src/in_memory/chains.rs [hunk 4]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+    // might need to use a map type and pop / reinsert with cummulative work as the index
+    chains: Vec<VolatileChain>,
+}
+
+impl<S> ChainsState<S> {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) -> Result<BlockHeaderHash, Error> {
+        let block = block.into();
+        let hash = block.hash();
+        let height = block.coinbase_height().unwrap();
+        let parent_height = BlockHeight(height.0 - 1);
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: When we insert the genesis block, this subtraction will panic (in debug mode), or produce an invalid height (in release mode).

File: zebra-state/src/in_memory/chains.rs [hunk 5]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+
+        for chain in self
+            .chains
+            .iter_mut()
+            .filter(|chain| chain.contains(parent_height))
+        {
+            let parent_state = chain
+                .0
+                .get(&parent_height)
+                .expect("block with one less height must exist");
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: This line will panic when:
* the state is empty, and we insert the genesis block.

If I understand the design correctly, it could also panic when:
* there are multiple chains of unequal length,
* we are trying to extend the tip of a chain, and
* a shorter chain is earlier in the list of chains.

I think that can't happen, because the map is ordered by height. If that's the case, let's document this invariant in the expect(), and in the data structure doc comment

We will also have to be careful of another edge case, where a shorter chain has a higher proof of work. Let's make sure we add tests for that.

File: zebra-state/src/in_memory/chains.rs [hunk 6]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+            let parent_hash = parent_state.block.hash();
+
+            if parent_hash != block.header.previous_block_hash {
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: This check duplicates a BlockVerifier check in zebra-consensus, see #737. zebra-consensus verifies the chain structure, because the chain structure is part of contextual verification.

Instead of re-verifying the chain structure here, the state could accept a request to add a block:
* as a child of a particular block hash,
* at a particular height,
* if the state satisfies a particular predicate (represented by a closure).

For heights, the predicate would be simple:
* the parent height is one less than the child height, or
* the block is a genesis block, the hash matches the genesis hash for the network, and the state is empty.

I think this design would help us keep the consensus rules in zebra-consensus.

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs]: As an interim design, we could let the BlockVerifier do all the checks itself, rather than passing a closure. Unfortunately, it would need to check the conditions in a delay loop, which increases CPU usage. For example, see #748.

**Review by teor2345 — CHANGES_REQUESTED** [review]
I also think there are a few panics in the code.

Can we just do the chain structure checks in the BlockVerifier instead?
Then we don't have to get the edge cases right in two places.

File: zebra-state/src/in_memory/chains.rs [hunk 7]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs:17]: Do we actually need the block here?

If it's already stored in the inner state, then we're storing it twice, which could lead to desynchronisation bugs.

File: zebra-state/src/in_memory/chains.rs [hunk 8]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+    // TODO: add more state here
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs:18]: Some state we will probably want to store:

```suggestion
    // TODO: add more state here
    //  - SetMainTip: a u32 (?) containing the cumulative proof of work
    //  - block.header.time: an im vector of recent block times
    //  - block.header.bits: an im vector of recent block difficulties
    //  - light_client_hash: an im vector of all subtree roots (Heartwood only)
    //  - other state required to validate transactions (as im data structures)
```

See https://github.com/ZcashFoundation/zebra/issues/477 for a summary of the header fields.

File: zebra-state/src/in_memory/chains.rs (hunk 7 above)
Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs:17]: We could just store the block hash.

In general, we should split the fields we need out into im data structures, rather than keeping a reference to the entire block. It will be faster to access, and easier to code.

Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs:17]: I wasn't planning on storing the block in the state while it's within the reorg limit

File: zebra-state/src/in_memory/chains.rs [hunk 9]
```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
...
+                .0
+                .get(&parent_height)
+                .expect("block with one less height must exist");
+            let parent_hash = parent_state.block.hash();
+
+            if parent_hash != block.header.previous_block_hash {
+                continue;
+            }
+
+            let was_present = if chain.contains(height) {
```
Comment (teor2345) [inline zebra-state/src/in_memory/chains.rs]: I think there's a really subtle bug here, triggered by another bug in the syncer.

1. The syncer submits a duplicate block for verification
2. It passes verification, based on the content of the parent block's chain state
3. The state already has 2 alternate tips, one contains the duplicate block, and one does not

If the tip that contains the duplicate block appears earlier in chains, the state layer panics.

If the tip that contains the parent block, but not the duplicate block, appears earlier in chains, the state layer creates a duplicate entry in chains. So now there are two entries in chains that contain exactly the same blocks.

Here's a different algorithm that ensures we don't end up with duplicate chains:
1. Store the ChainList as a HashMap by block hash. Each Chain in the ChainList is stored as an OrdMap by block height.
2. When the state gets a new block, check for its hash in the ChainList. If it exists, error or panic. (Probably error for now, until we fix the syncer.)
3. If the new block does not exist, check for its parent block in ChainList. If the parent not exist, error or panic. (Error until #737 is merged, then panic.)
4. Take a copy of the parent block's Chain, and insert the new block as a child of the parent block. Add the new Chain to the ChainList.

File: zebra-state/src/in_memory/chains.rs (hunk 8 above)
Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs:18]: I'd prefer to collapse these bits of information into the struct and then use `im` to track a vector of the structs, rather than tracking multiple vectors of fields. That way only the block time for that block will be stored in each "ChainState" but the im vector of chain states will give you access to all the recent block times, difficulties, etc.

File: zebra-state/src/in_memory/chains.rs (hunk 5 above)
Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs]: I don't think this depends on the map being ordered, we can't panic here because we filtered out chains that don't contain an element at `parent_height`, and then we retrieve that element. Come to think of it I can re-write this in a way that doesn't even panic, though I bet in release mode rust optimizes out the panic as is.

File: zebra-state/src/in_memory/chains.rs (hunk 7 above)
Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs:17]: I'm not sure what the best design is here.

If we store the blocks outside the chain context, then we could accidentally drop a block, but keep its verification context.

If we store the blocks in the chain context, then the data we use for verification is less explicit. There could also be a slightly higher cost to access recent blocks, rather than a specific data structure. But I don't think it will matter in practice.

I think storing data in multiple places is actually a greater risk.

File: zebra-state/src/in_memory/chains.rs (hunk 8 above)
Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs:18]: I think you're right about `time` and `bits`. I don't want to duplicate data in recent blocks and in the context.

I think the `light_client_hash` set is different, because it could come from any Heartwood block (including those on disk). But the actual number of roots is very small. So I think the most efficient structure is an `im::Vector` of roots.

We should persist a single set of roots to disk, every time we persist a block. Then we can initialise the root set from disk when we load the verifier.

For more details, see the Heartwood section of:
https://github.com/ZcashFoundation/zebra/pull/742#discussion_r460380882

Comment (teor2345) [reply zebra-state/src/in_memory/chains.rs:18]: Similarly, I think the most efficient way to work with the large UTXO and nullifier sets is an `im` map of deltas.

If we don't use `im` data structures, we could end up using a lot of memory by duplicating parts of these sets in each context.

Comment (yaahc) [reply zebra-state/src/in_memory/chains.rs:18]: :thumbsup:

**Comment by teor2345** [comment]
I wrote up a draft RFC for parallel verification in #763.

I think it resolves a bunch of design issues in this PR, let me know what you think.

File: zebra-state/src/on_disk.rs [hunk 10]
```diff
@@ -268,7 +251,7 @@ type BoxError = Box<dyn error::Error + Send + Sync + 'static>;
 // Please forgive me.
 
 /// a type that can store any error and implements the Error trait at the cost of
-/// not implementing From<E: Error>
+/// not implemnting From<E: Error>
```
Comment (yaahc) [inline zebra-state/src/on_disk.rs]:
```suggestion
not implementing From<E: Error>
```

File: zebra-state/src/on_disk.rs [hunk 11]
```diff
@@ -268,7 +251,7 @@ type BoxError = Box<dyn error::Error + Send + Sync + 'static>;
 // Please forgive me.
 
 /// a type that can store any error and implements the Error trait at the cost of
-/// not implementing From<E: Error>
+not implementing From<E: Error>
```
Comment (yaahc) [inline zebra-state/src/on_disk.rs]:
```suggestion
/// not implementing From<E: Error>
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for pulling this together!

Can you add a detailed module comment?
It should explain each struct, its purpose, how it is used, and how they relate to one another.

This PR adds `Chain`, `ChainState`, `ChainsState` (notice the extra "s"), `ChainSet`, and `ChainContext` structs.
I find it hard to understand this code, because those names are so similar.

Can you rename them so the names are distinct, and fit what they are, and how they are used?

(After that, I'd like to re-review, because I think we might have 1-2 more structs than we need.)
